### PR TITLE
API: remove deep keyword from EA.copy

### LINKED ISSUE #26995: Non-Deep copy for Categorical?
Categorical.copy passes `values=self._codes.copy()`, and doesn't have a `deep=False` kwarg (which the base ExtensionArray does).  This is causing trouble in #26954.  Is this intentional @TomAugspurger?

Pandas-internal EAs with `copy` behavior not respecting `deep` kwarg:
- [ ] Categorical (no kwarg at all)
- [ ] DatetimeLike (ignores kwarg)
- [ ] PandasArray

IntegerArray, IntervalArray, and SparseArray look correct.

### PR DESCRIPTION
**Pull request by jbrockmendel**
Not clear on what type of deprecation cycle we want here

- [x] closes #26995, closes #27024

### CODE REVIEW

File: pandas/tests/extension/decimal/array.py [hunk 1]
```diff
@@ -101,10 +101,8 @@ def take(self, indexer, allow_fill=False, fill_value=None):
                       allow_fill=allow_fill)
         return self._from_sequence(result)
 
-    def copy(self, deep=False):
```
Comment (jreback) [inline pandas/tests/extension/decimal/array.py]: can you add a test in the pandas/tests/extension/constructors.py so it hits all EA's (for view & copy)

**Review by jreback — CHANGES_REQUESTED** [review]
I would be ok with just making this change (no deprecation cycle), @jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
cc @TomAugspurger 

Looks good to me. As usual (sorry ...), I would only add `view()` if we actually need it (meaning: use it internally)

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -820,19 +820,30 @@ def take(self, indices, allow_fill=False, fill_value=None):
         # pandas.api.extensions.take
         raise AbstractMethodError(self)
 
-    def copy(self, deep: bool = False) -> ABCExtensionArray:
+    def copy(self) -> ABCExtensionArray:
```
Comment (TomAugspurger) [inline pandas/core/arrays/base.py]: We need to deprecate `deep` right? Or is in not in 0.24?

Comment (TomAugspurger) [reply pandas/core/arrays/base.py]: Oh, just saw https://github.com/pandas-dev/pandas/pull/27083#pullrequestreview-255410112.

Slight preference for deprecation, but don't have strong thoughts here.

File: pandas/core/arrays/base.py [hunk 3]
```diff
@@ -820,19 +820,30 @@ def take(self, indices, allow_fill=False, fill_value=None):
...
         """
         Return a copy of the array.
 
+        Returns
+        -------
+        ExtensionArray
+        """
+        raise AbstractMethodError(self)
+
+    def view(self, dtype=None) -> ABCExtensionArray:
```
Comment (jreback) [inline pandas/core/arrays/base.py]: ok with .view() @jorisvandenbossche @TomAugspurger ?

**Review by jreback — CHANGES_REQUESTED** [review]
@jbrockmendel can you add a whatsnew note that remove the deep kwarg for EA

**Comment by jorisvandenbossche** [comment]
> Slight preference for deprecation, but don't have strong thoughts here.

So a user only runs into this when they directly interface with an ExtensionArray themselves (as the uses within pandas get fixed). Further, for EA authors only implementing the EA itself, it's not a backwards incompatible change, as we don't have a problem that the EA still has the keyword, it will just never be called from inside pandas.

It seems that users doing `Series.array.copy(deep=True)` is something very specific, and also only introduced very recently (eg Categorical is longest out there, but has actually no copy keyword). IntegerArray and IntervalArray eg does have one.

File: pandas/core/arrays/base.py (hunk 3 above)
Comment (jreback) [reply pandas/core/arrays/base.py]: actually let's remove this for now @jbrockmendel

**Comment by TomAugspurger** [comment]
Sounds good.

File: pandas/core/arrays/sparse.py [hunk 4]
```diff
@@ -1262,12 +1262,8 @@ def searchsorted(self, v, side="left", sorter=None):
...
 
-    def copy(self, deep=False):
-        if deep:
-            values = self.sp_values.copy()
-        else:
-            values = self.sp_values
-
+    def copy(self):
+        values = self.sp_values.copy()
         return self._simple_new(values, self.sp_index, self.dtype)
```
Comment (jbrockmendel) [inline pandas/core/arrays/sparse.py]: @TomAugspurger do we need to copy the index here?

Comment (TomAugspurger) [reply pandas/core/arrays/sparse.py]: Not really sure. I don't know if they're designed for sharing across instances, or whether we mutate them inplace.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

File: pandas/core/arrays/sparse.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/sparse.py]: In any case in sparse.py, we don't mutate them inplace. I would also assume it is fine not to copy them.

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**